bootloader/grub.go

```diff
@@ -525,65 +573,87 @@ func (g *grub) getGrubBootAssetsForArch() (*grubBootAssetPath, error) {
 	if err != nil {
 		return nil, err
 	}
-	return []string{assets.shimBinary, assets.grubBinary}, nil
+	return [][]taggedPath{{assets.shimBinary, assets.grubBinary}, {assets.defaultShimBinary, assets.defaultGrubBinary}}, nil
 }
 
-// getGrubRunModeTrustedAssets returns the assets for run mode, which is
-// grub from the boot partition.
-func (g *grub) getGrubRunModeTrustedAssets() ([]string, error) {
+// getGrubRunModeTrustedAssets returns the list of ordered asset
+// chains for run mode, which is grub from the boot partition.
+func (g *grub) getGrubRunModeTrustedAssets() ([][]taggedPath, error) {
 	assets, err := g.getGrubBootAssetsForArch()
 	if err != nil {
 		return nil, err
 	}
-	return []string{assets.grubBinary}, nil
+	return [][]taggedPath{{assets.defaultGrubBinary}}, nil
 }
 
-// TrustedAssets returns the list of relative paths to assets inside
-// the bootloader's rootdir that are measured in the boot process in the
-// order of loading during the boot.
-func (g *grub) TrustedAssets() ([]string, error) {
+// TrustedAssets returns the map of relative paths to asset
+// identifers. The relative paths are relative to the bootloader's
+// rootdir. The asset identifiers correspond to the names of BootFile
+// given by BootChains.
+func (g *grub) TrustedAssets() (map[string]string, error) {
 	if !g.nativePartitionLayout {
 		return nil, fmt.Errorf("internal error: trusted assets called without native host-partition layout")
 	}
+	ret := make(map[string]string)
+	var chains [][]taggedPath
+	var err error
 	if g.recovery {
-		return g.getGrubRecoveryModeTrustedAssets()
+		chains, err = g.getGrubRecoveryModeTrustedAssets()
+	} else {
+		chains, err = g.getGrubRunModeTrustedAssets()
+	}
+	if err != nil {
+		return nil, err
+	}
+	for _, chain := range chains {
+		for _, asset := range chain {
+			ret[asset.path] = asset.GetId()
+		}
 	}
-	return g.getGrubRunModeTrustedAssets()
+	return ret, nil
 }
 
-// RecoveryBootChain returns the load chain for recovery modes.
+// RecoveryBootChains returns the list of load chains for recovery modes.
 // It should be called on a RoleRecovery bootloader.
-func (g *grub) RecoveryBootChain(kernelPath string) ([]BootFile, error) {
+func (g *grub) RecoveryBootChains(kernelPath string) ([][]BootFile, error) {
 	if !g.recovery {
 		return nil, fmt.Errorf("not a recovery bootloader")
 	}
 
 	// add trusted assets to the recovery chain
-	assets, err := g.getGrubRecoveryModeTrustedAssets()
+	assetsSet, err := g.getGrubRecoveryModeTrustedAssets()
 	if err != nil {
 		return nil, err
 	}
-	chain := make([]BootFile, 0, len(assets)+1)
-	for _, ta := range assets {
-		chain = append(chain, NewBootFile("", ta, RoleRecovery))
+	chains := make([][]BootFile, 0, len(assetsSet))
+	for _, assets := range assetsSet {
+		chain := make([]BootFile, 0, len(assets)+1)
+		for _, ta := range assets {
+			// We use identifiers as paths. They will be
+			// resolved in buildBootAssets from
+			// boot/seal.go from the modeenv
```

Comment:
this is a bit strange given that BootFile has a still a field called Path, I thought we would keep the path as they were and use the TrustedAssets map to get to id to use in the cache

Reply:
Thank you for the tip. I did not realize I could do that. I have fixed it.